include/swift/Runtime/Concurrency.h

```diff
@@ -627,6 +627,17 @@ void swift_task_switch(SWIFT_ASYNC_CONTEXT AsyncContext *resumeToContext,
                        TaskContinuationFunction *resumeFunction,
                        ExecutorRef newExecutor);
 
+/// Mark a task for enqueue on a new executor and then enqueue it.
+///
+/// The resumption function pointer and continuation should be set
+/// appropriately in the task.
+///
+/// Generally you should call swift_task_switch to switch execution
+/// synchronously when possible.
+SWIFT_EXPORT_FROM(swift_Concurrency) SWIFT_CC(swift)
```

Comment:
@rjmccall, with the changes I made to provide a new `flagAsEnqueuedOnExecutor` on the task, which does the necessary bookkeeping and enqueuing of the task, these unit tests started failing since they were directly calling `swift_task_enqueue` without doing the right bookkeeping ahead of time.

I therefore had to introduce a new exported function here, for the sole purpose of being able to get the unit tests working and exercising the logic here. It is not needed for the actual logic to work. And on that note, `swift_task_enqueue` and `swift_task_enqueueGlobal` should also not to be exported since they are only internal to the runtime library.

How do you want to proceed here?

Reply:
Hmm.  I hadn't really thought about the public entrypoints.  They're not usually directly useful, but:

- We do have a way to create a job in an unscheduled state, and we ought to have *some* way to schedule the task afterwards.  That could be a new operation.  I guess it depends on whether you actually found it useful to do the bookkeeping in the caller, or whether it's necessary to touch the status during enqueue regardless?  I need to look at the new patch.

- We also ought to continue to provide some way to hook the scheduling process, in case we need that for future compatibility hackery.  So at any rate, we shouldn't remove the ability to replace these.

Reply:
Is there a way to set up a created-but-not-scheduled task's status so that enqueuing it just works, or do we need to add specific new API for this?

Reply:
> We do have a way to create a job in an unscheduled state, and we ought to have some way to schedule the task afterwards. That could be a new operation. I guess it depends on whether you actually found it useful to do the bookkeeping in the caller, or whether it's necessary to touch the status during enqueue regardless? I need to look at the new patch.

Is "create a job but don't enqueue it" currently supported in the language? I agree that it is a useful thing to support in the future. In my current patch, `flagAsEnqueuedOnExecutor` handles 2 possible transitions of a task to the enqueued state: `suspended->enqueued` and `running->enqueued`. When a task is created but not scheduled, that falls under the first transition that is handled in `flagAsEnqueuedOnExecutor`.

I think both codepaths - whether the task is coming from suspended or from running - will need to touch the status to at least set the enqueued bit on the task and sometimes need to touch the status in the caller. We may be able to refactor this into `flagAsEnqueuedFromRunning()` and `flagAsEnqueuedFromSuspended` in order to keep all the atomic operations on the status together. 

> Is there a way to set up a created-but-not-scheduled task's status so that enqueuing it just works, or do we need to add specific new API for this?

Yes and that's what we do in `swift_task_create_commonImpl`. The task is created earlier and then we call flagAsEnqueuedOnExecutor and enqueueing just works.

Reply:
> Is "create a job but don't enqueue it" currently supported in the language?

There's not a way to do it, no, but that's not the point; if we add an optimization / new feature that would use this facility, the question is whether it can back-deploy.  It's currently tentatively supported by the ABI; if that support isn't really enough to implement a reasonable future feature, so be it, we can say that that feature doesn't back-deploy; but it'd be nice to at least think about this.

> Yes and that's what we do in swift_task_create_commonImpl

No, I mean in the path where `taskCreateFlags.enqueueJob()` is not true.

Reply:
Ah I see where you're going with this - I didn't consider back deployment. 

There is currently nothing special in the status of a task prior to enqueueing - it just needs to not already have an enqueued bit. 

Assuming we consider "create a schedule-able task to enqueue later" as being in 3 parts: 
1. create a suspended task
2. (Optionally) Update a suspended task with any state changes if needed
3. Enqueue the suspended task when it is ready

All 3 parts are doable with the hook points we have today:
(1) with `taskCreateFlags.enqueueJob() == false`, 
(2) This is handled by the various operations that may need to touch the task like escalation and cancellation
(3) We can enqueue the suspended task with `flagAsEnqueueOnExecutor`. 

I don't think we need new API for it.